tests/backup/backup_upgrade_test.go

```diff
@@ -204,3 +205,366 @@ var _ = Describe("{StorkUpgradeWithBackup}", func() {
+				log.InfoD("Creating a backup of namespace [%s] without pre- and post-backup rules", namespace)
+				err := CreateBackup(backupName, SourceClusterName, backupLocationName, backupLocationUid, []string{namespace},
+					labelSelectors, orgID, srcClusterUid, "", "", "", "", ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of backup [%s]", backupName))
+				backupWithoutRuleNames = SafeAppend(&mutex, backupWithoutRuleNames, backupName).([]string)
+			}
+			_ = TaskHandler(Inst().AppList, createBackupWithoutRulesTask, Parallel)
+		})
+		Step("Create a schedule policy", func() {
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			intervalInMins = 5
+			log.InfoD("Creating a schedule policy with interval [%v] mins", intervalInMins)
+			schedulePolicyName = fmt.Sprintf("interval-%v-%v", intervalInMins, time.Now().Unix())
+			schedulePolicyInfo := Inst().Backup.CreateIntervalSchedulePolicy(5, int64(intervalInMins), 5)
+			err = Inst().Backup.BackupSchedulePolicy(schedulePolicyName, uuid.New(), orgID, schedulePolicyInfo)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of schedule policy [%s] with interval [%v] mins", schedulePolicyName, intervalInMins))
+			schedulePolicyUid, err = Inst().Backup.GetSchedulePolicyUid(orgID, ctx, schedulePolicyName)
+			log.FailOnError(err, "Fetching uid of schedule policy [%s]", schedulePolicyName)
+			log.Infof("Schedule policy [%s] uid: [%s]", schedulePolicyName, schedulePolicyUid)
+		})
+		Step("Create schedule backup for each namespace", func() {
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			createSingleNSBackupTask := func(appName string) {
+				namespace := appNamespaces[appName][0]
+				labelSelectors := make(map[string]string)
+				singleNSScheduleName := fmt.Sprintf("%s-%s-single-ns-schedule-%v", BackupNamePrefix, namespace, time.Now().Unix())
+				log.InfoD("Creating schedule backup with schedule [%s] of namespace [%s]", singleNSScheduleName, namespace)
+				err = CreateScheduleBackup(singleNSScheduleName, SourceClusterName, backupLocationName, backupLocationUid, []string{namespace},
+					labelSelectors, orgID, preRuleNames[appName], preRuleUids[appName], postRuleNames[appName], postRuleUids[appName], schedulePolicyName, schedulePolicyUid, ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of schedule backup with schedule [%s]", singleNSScheduleName))
+				firstScheduleBackupName, err := GetFirstScheduleBackupName(ctx, singleNSScheduleName, orgID)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching name of the first schedule backup with schedule [%s]", singleNSScheduleName))
+				log.Infof("First schedule backup name: [%s]", firstScheduleBackupName)
+				singleNamespaces = SafeAppend(&mutex, singleNamespaces, namespace).([]string)
+				singleNSScheduleNames = SafeAppend(&mutex, singleNSScheduleNames, singleNSScheduleName).([]string)
```

Comment:
This will be faster with separate locks, as each lock would be protecting a different thing,
use separate mutex for `singleNamespaces`, as lock should "belong" to the object that it is protecting.
use separate mutex for `singleNSScheduleNames`, as lock should "belong" to the object that it is protecting.